modules/http/src/test/java/org/mule/module/http/functional/listener/HttpListenerAttachmentsTestCase.java

```diff
@@ -112,6 +122,47 @@ public void receiveOnlyAttachmentsAndReturnOnlyAttachmentsWithMultipartFormDataA
     }
 
     @Test
+    public void receiveAttachmentWithCustomHeaders() throws Exception
+    {
+
+        RequestBuilder requestBuilder = new RequestBuilder();
+        AsyncHttpClientConfig asyncHttpClientConfig = new AsyncHttpClientConfig.Builder().build();
+        GrizzlyAsyncHttpProvider grizzlyAsyncHttpProvider = new GrizzlyAsyncHttpProvider(asyncHttpClientConfig);
+
+        requestBuilder.setMethod("POST");
+        requestBuilder.setUrl(getUrl(formDataPath.getValue()));
+        ByteArrayPart byteArrayPart = new ByteArrayPart("partName", "blahblah".getBytes(), "application/pdf", null, null, null, "base64");
+        byteArrayPart.setCustomContentDisposition("file; name=\"partName\"; documentId=1");
+        byteArrayPart.addCustomHeader("Custom-Header: ", "custom");
+        requestBuilder.addBodyPart(byteArrayPart);
+
+        try (AsyncHttpClient asyncHttpClient = new AsyncHttpClient(grizzlyAsyncHttpProvider, asyncHttpClientConfig))
+        {
+
+            ListenableFuture<Response> responseFuture = asyncHttpClient.executeRequest(requestBuilder.build());
+            com.ning.http.client.Response response = responseFuture.get();
+
+            // check that the Mule message is created correctly from the HTTP request
+            final MuleMessage receivedMessage = muleContext.getClient().request(VM_MESSAGE_ENDPOINT, 1000);
```

Comment:
No VM transport. You'll need to use a <test:queue> in 4.x. Just remember it for the cherry-pick.